g2o/core/base_constant_edge.h

```diff
@@ -0,0 +1,185 @@
+// This code is based on g2o/core/base_binary_edge.h and
+// g2o/core/base_multi_edge.h
+//
+// g2o - General Graph Optimization
+// Copyright (C) 2011 R. Kuemmerle, G. Grisetti, H. Strasdat, W. Burgard
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright
+//   notice, this list of conditions and the following disclaimer in the
+//   documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef G2O_BASE_CONSTANT_EDGE_H
+#define G2O_BASE_CONSTANT_EDGE_H
+
+
+#include <iostream>
+#include <limits>
+
+#include "base_edge.h"
+#include "robust_kernel.h"
+#include "g2o/config.h"
+#include "g2o/stuff/misc.h"
+
+namespace g2o {
+
+  namespace {
+    // assumes i < j
+    constexpr int pair_to_index(const int i, const int j)
+    {
+      return j * (j - 1) / 2 + i;
+    }
+    constexpr std::pair<int, int> index_to_pair(const int k)
+    {
+      const int j = std::sqrt(2*(k+1))+.5;
+      const int i = k - j * (j - 1) / 2;
+      return {i, j};
+    }
+
+    template<int N, typename... Types>
+    using NthType = typename std::tuple_element<N, std::tuple<Types...>>::type;
+
+    template<typename Value, typename... Ts2>
+    auto tuple_init(const Value& value, const std::tuple<Ts2...>)
```

Comment:
Since you're just using this to assign all the values in your tuple to nullptr it's redundant. All you need is to invoke the default constructor {}, and that will do it for you. You could even just initialize it at the point of declaration which would simplify this as well.

Reply:
Ah I see these aren't pointers you're trying to call the nullptr constructor of that type. It's annoying that they aren't default constructible that would have simplified things.

Reply:
If you defined a NullType you could do it like this:

template<typename Type>
using NullType = nullptr_t;

      BaseConstantEdge() : BaseEdge<D,E>(),
      _hessianRowMajor(false),
      _hessianTuple{NullType<VertexTypes>{}...},
      _hessianTupleTransposed{NullType<VertexTypes>{}...},
      _jacobianOplus({nullptr, D, VertexTypes::Dimension}...)
      {
        _vertices.resize(_nr_of_vertices, nullptr);
      }

Reply:
Ah, nice one. However gcc 6.3 gives strange compilation errors when I changed it like that. It starts complaining about the _jacobianOplus initialization, though that line is not changed. Do you know why this is?
<hmmm, I can't seem to attach a log of the error, will try later>

Reply:
Ah it's because in the other version you're expanding to nr_of_vertex_pairs not VertexTypes, I'll have to look it over again

Reply:
Ah, of course, I overlooked that. Well, by moving the trick inside struct HessianTupleType I can make it to work:
```
      struct HessianTupleType<std::integer_sequence<std::size_t, Ints...>>
      { 
        using type = std::tuple<HessianBlockTypeK<Ints>...>;
        using typeTransposed = std::tuple<HessianBlockTypeKTransposed<Ints>...>;
        template<std::size_t I>
        using NullType = nullptr_t; 
        static auto init() { return type{NullType<Ints>{}...};}
        static auto initTransposed() { return typeTransposed{NullType<Ints>{}...};}
      };
      using HessianTupleTypeH = HessianTupleType<std::make_index_sequence<_nr_of_vertex_pairs>>;
      BaseConstantEdge() : BaseEdge<D,E>(),
      _hessianRowMajor(false),
      _hessianTuple(HessianTupleTypeH::init()),
      _hessianTupleTransposed(HessianTupleTypeH::initTransposed()),
      _jacobianOplus({nullptr, D, VertexTypes::Dimension}...)
```
But that looks actually more complicated than the tuple_init function...

Reply:
Yeah I think you're right, do what you think is best. You may want to change that tuple init not to grab that tuple by value (const&). Or just use tuple_init<decltype(_hessionTuple)> instead of the parameter to do the capture of the tuple type.

Reply:
Changed it into a const&.